TST Add minimal setup to be able to run test suite on float32

jjerphan opened the pull request:
#### Reference Issues/PRs

Solves #22680.

#### What does this implement/fix? Explain your changes.

Based on @thomasjpfan's https://github.com/scikit-learn/scikit-learn/pull/22663#issuecomment-1058266001, this implements a minimal setup to run the test suite on a `dtype` parameter using a minimal parametrisation.

#### Any other comments?

jjerphan committed:
TST Add minimal setup to be able to run test suite on float32

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

ogrisel commented:
To avoid merging dead code could you please use it in at least one test and set the environment variable on one of the CI builds (a fast one preferably)?

jjerphan commented:
Yes, this is still in draft and I am figuring which tests I am going to include here based on the discussions in other pull requests.

jjerphan committed:
TST Use dtype fixture in one test

jjerphan committed:
CI Do not skip 32bit test for py38_conda_defaults_openblas

jjerphan commented on sklearn/conftest.py:
```diff
@@ -37,6 +38,17 @@
     "fetch_rcv1_fxt": fetch_rcv1,
 }
 
+_SKIP32_MARK = pytest.mark.skipif(
+    environ.get("SKLEARN_SKIP_FLOAT32", "1") != "0",
+    reason="Set SKLEARN_SKIP_FLOAT32=0 to run float32 dtype tests",
+)
```
How about using this phrasing and semantic?
```suggestion
_SKIP32_MARK = pytest.mark.skipif(
    environ.get("SKLEARN_RUN_FLOAT32", "0") != "1",
    reason="Set SKLEARN_RUN_FLOAT32=1 to run float32 dtype tests",
)
```

I find it more intuitive. What do you think?

thomasjpfan replied:
I like you suggestion. In general, I think the "positive" version with "SKLEARN_RUN_*` is easier to reason about.

jjerphan committed:
Merge branch 'main' into float32-test-suite

jjerphan committed:
fixup! TST Use dtype fixture in one test

jjerphan commented:
I am just adding a test for this global fixture.
I would prefer updating the set of recent `TST` PRs accordingly if this PR gets approved.

thomasjpfan commented on sklearn/conftest.py:
```diff
@@ -37,6 +38,17 @@
+
+# Global fixture
+@pytest.fixture(params=[pytest.param(np.float32, marks=_SKIP32_MARK), np.float64])
+def dtype(request):
```
We use "dtype" as a `pytest.mark.parameterize" in many places where we actually want to test for all dtypes. Maybe we need to give this a more semantic name:

```suggestion
def global_dtype(request):
```

I am open for other suggestions.

thomasjpfan commented on sklearn/conftest.py:
```diff
@@ -37,6 +38,17 @@
+
+# Global fixture
+@pytest.fixture(params=[pytest.param(np.float32, marks=_SKIP32_MARK), np.float64])
+def dtype(request):
```
We use "dtype" as a `pytest.mark.parameterize" in many places where we actually want to test for all dtypes. Maybe we need to give this a more semantic name:

```suggestion
def global_dtype(request):
```

I am open for other suggestions.

jjerphan replied:
I agree that we need to make explicit that the current `dtype` is a global fixture. To me `pytest.fixture` comes in handy but with implicit definition which might be misleading for new comers reading the tests.

Could we come up with a convention for fixture definition and usage?

Though, I can not imagine a good option to make fixture explicit, I am OK using your proposed `global_` prefix as a convention for now.

jjerphan replied:
Addressed in 09ccf3196c948532974fb40d67a1f14813f7b06d.

jjerphan committed:
MAINT Apply reviews comments

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

ogrisel commented on azure-pipelines.yml:
```diff
@@ -201,6 +201,7 @@ jobs:
         MATPLOTLIB_VERSION: 'min'
         THREADPOOLCTL_VERSION: '2.2.0'
         SKLEARN_ENABLE_DEBUG_CYTHON_DIRECTIVES: '1'
+        SKLEARN_RUN_FLOAT32: '1'
```
Let's make it explicit that this new environment variable only impacts the tests:

```suggestion
        SKLEARN_TESTS_RUN_FLOAT32: '1'
```

ogrisel commented on sklearn/conftest.py:
```diff
@@ -37,6 +38,17 @@
 }
 
+_SKIP32_MARK = pytest.mark.skipif(
+    environ.get("SKLEARN_RUN_FLOAT32", "0") != "1",
+    reason="Set SKLEARN_RUN_FLOAT32=1 to run float32 dtype tests",
```
```suggestion
    environ.get("SKLEARN_TESTS_RUN_FLOAT32", "0") != "1",
    reason="Set SKLEARN_TESTS_RUN_FLOAT32=1 to run float32 dtype tests",
```

ogrisel reviewed (approved):
If we start to use this fixture extensively, this will generate a lot of skipped tests by default which will render the output very verbose. At some point we might want to avoid the skip mark and instead do:

```python
GLOBAL_DTYPES_TO_TEST = [np.float64]
if environ.get("SKLEARN_TESTS_RUN_FLOAT32", "0") != "1":
    print("Enabling float32 tests globally.")
    GLOBAL_DTYPES_TO_TEST.append(np.float32)
else:
    print("Skipping float32 tests globally. Set SKLEARN_TESTS_RUN_FLOAT32=1 to enable.")


@pytest.fixture(params=GLOBAL_DTYPES_TO_TEST)
def global_dtype(request):
    yield request.param
```

but I think we can keep it this way for now. +1 for merging as is.

ogrisel commented:
I have an idea to reuse a similar pattern to define a `random_seed` fixture.

jjerphan committed:
Use an more explicit name for the env variable

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

ogrisel commented:
@thomasjpfan let's merge?

thomasjpfan reviewed (commented):
Codewise this looks good. Can we document the variable along with the other ones here?

https://github.com/scikit-learn/scikit-learn/blob/05ef75200f5216d2a30907aed5fcb41c50b7cc78/doc/computing/parallelism.rst?plain=1#L203-L207

(A future PR can be to move some of these environment variables out of `doc/computing/parallelism.rst` and into `doc/developers/contributing.rst`.)

jjerphan committed:
DOC Rename and document the environement variable

So as to have a similar name to `SKLEARN_SKIP_NETWORK_TESTS`.

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan committed:
fixup! DOC Rename and document the environement variable

jeremiedbb reviewed (dismissed):
I think this fixture is not enough on its own

thomasjpfan commented:
In PyTorch, sets their own tolerances based on dtype: https://github.com/pytorch/pytorch/blob/1f29b3130af218847a043e58fdc64511bbe072fe/torch/testing/_comparison.py#L43-L51 (They use `unittest`, so they can use `self.assertEqual` that sets the tolerances properly)

For our use case, we can have a custom "assert_allclose" that sets the `atol` and `rtol` correctly based on the dtypes.

jeremiedbb commented:
I'd be happy with a custom `assert_allclose`. To me it should be done in this PR or prior to this PR before we start getting a bunch of PRs that add the fixture with random fixes to make the tests pass.

ogrisel commented:
+1 as well for a custom `assert_allclose` in scikit-learn.

jjerphan commented:
I also think having custom assertions (e.g `assert_allclose`) for testing would be convenient.

Currently, assertions in tests  are imported from `numpy.testing` and `sklearn._utils.testing`.
Moreover, `sklearn._utils.testing` simply declare its assertions by importing the ones `numpy.testing`.

In this regards,  I think that it would be better to import assertions from `sklearn._utils.testing` instead of from `numpy.testing`, prior the definition of custom versions of assertions.

What do you think?

jeremiedbb commented:
To me the order should be:
- Define a custom assert_allclose
- Define the fixture (can be at the same time as first item)
- progressively add the fixture to wisely selected tests. When adding the fixture to a test, make sure that we use the custom assert_allclose, and check that the rtol and atol are meaningful.

jjerphan committed:
Merge branch 'main' into float32-test-suite

jjerphan committed:
TST Introduce custom assert_allclose

This is heavily inspired by the one in PyTorch:
https://github.com/pytorch/pytorch/blob/1f29b3130af218847a043e58fdc64511bbe072fe/torch/testing/_comparison.py

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

jjerphan commented:
Done.

I meant, the current interfaces for tests assertions must be sorted as well because assertions might not always be routed to the proper ones (the ones defined in `sklearn._utils.testing`, I suppose).

ogrisel reviewed (commented):
The generic idea of the new `assert_allclose` LGTM but it requires its own testing:

jeremiedbb commented on sklearn/utils/_testing.py:
```diff
@@ -87,6 +88,13 @@
 
+DTYPE_TOLERANCES = {
+    #            rtol, atol
+    np.float32: (1e-6, 1e-5),
+    np.float64: (1e-7, 1e-7),
```
I think 1e-6 is way too ambitious. We have many tests that won't pass with that. I think it should be 1e-4. (we are not testing setting 1e-15 for float64 afterall).

I think we should have atol=0 by default. atol only make sense to compare against zero. In addition it represents what value can be considered to be zero which depends on the magnitude of the non zeros entries of the array.

ogrisel replied:
> I think we should have atol=0 by default. atol only make sense to compare against zero. In addition it represents what value can be considered to be zero which depends on the magnitude of the non zeros entries of the array.

But then this will make it harder to write atol tests that are dtype aware, no? Do you expect test writers to do something like the following?

```python
# b has some zero entries
assert_allclose(a, b, atol=1e-4 if dtype == np.float32 else 1e-7)
```

jeremiedbb replied:
Since atol should only be used in very specific cases, I think it's acceptable.

Having a non zero default for atol, let's say 1e-7, leads to:
```py
assert_allclose(1e-7, 2e-7)
# no error
```
It's equivalent to using `assert_array_almost_equal` which is what I (we ?) fight against since I joined the project.
non-zero values should be compared using relative difference. Comparing against zero requires absolute difference.

ogrisel replied:
Alright, thanks for the clarification. I think once this PR is merged, we should write guidelines on how to properly use `assert_allclose` in our test suite.

At the moment we do not mention it in our doc: https://scikit-learn.org/dev/search.html?q=assert_allclose

jjerphan replied:
0d42609 added tests, adapted the relative tolerances, and set the absolute tolerance to 0 by default.

jjerphan committed:
Review comments

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
TST Add tests to testing tests

Yes.

jjerphan committed:
Merge branch 'main' into float32-test-suite

jjerphan committed:
TST Add more rtols

jjerphan committed:
TST Adapt test_testing.py

jeremiedbb reviewed (commented):
Looking at what we have to do to implement our custom assert_allclose and the fact that we won't have a dtype dependent atol, I wonder if it wouldn't be simpler to make the fixture being `"global_dtype, rtol, atol"` ?

jjerphan commented on sklearn/utils/tests/test_testing.py:
```diff
@@ -854,3 +855,61 @@ def test_raises():
+    a = np.zeros(10, dtype=global_dtype) + dtype_eps
+    b = np.zeros(10, dtype=global_dtype)
+
+    msg = "Not equal to tolerance rtol=1e-09, atol=0"
```
It has been updated in recent commits.

jjerphan commented on sklearn/utils/_testing.py:
```diff
@@ -387,6 +397,72 @@ def assert_raise_message(exceptions, message, function, *args, **kwargs):
+    """
+    dtypes = []
+
+    for input in (actual, desired):
+        if _is_arraylike_not_scalar(input):
+            dtypes.append(np.asarray(input).dtype)
+        elif isinstance(input, np.dtype):
+            dtypes.append(input)
+        elif isinstance(input, Number):
+            dtypes.append(np.dtype(type(input)))
+        else:
+            raise TypeError(
+                "Expected a scalar, a np.array or a np.dtype, but got"
+                f" {type(input)} instead."
+            )
```
Way better.

ogrisel commented:
I suppose the deletion of `sklearn/conftest.py` was not intentional.

ogrisel commented:
> Looking at what we have to do to implement our custom assert_allclose and the fact that we won't have a dtype dependent atol, I wonder if it wouldn't be simpler to make the fixture being "global_dtype, rtol, atol" ?

I do not understand.

We need a fixture to enable or disable the float32 tests to avoid doubling the test duration by default no? Then if those dtype-parametrized tests use our `assert_allclose` the body of those test function will be dtype-agnostic almost magically, no? The only cases where this won't necessary be the case will be the tests where we expect zero values in which case choosing an atol carefully will have to be done on a case by case basis: https://github.com/scikit-learn/scikit-learn/pull/22690#discussion_r826077696

jjerphan committed:
Julien clearing his mess

jjerphan committed:
Simplify

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

jjerphan committed:
TST Trust numpy test suite

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

jjerphan committed:
C'mon, Julien

jjerphan commented on sklearn/utils/_testing.py:
```diff
@@ -387,6 +397,72 @@ def assert_raise_message(exceptions, message, function, *args, **kwargs):
 
 
+def assert_allclose(
+    actual, desired, rtol=None, atol=0.0, equal_nan=True, err_msg="", verbose=True
```
I just wanted to reflect the same interface to ease drop in replacement.

I am +0 to change it. WDTY, @ogrisel?

ogrisel replied:
I agree we should probably mirror numpy's public API for the other parameters to avoid surprises.

jeremiedbb replied:
You're right, let's keep it

jjerphan committed:
TST Actually use custom assert_allclose

jjerphan commented on sklearn/feature_selection/tests/test_mutual_info.py:
```diff
@@ -21,7 +21,7 @@ def test_compute_mi_dd():
 
 
-def test_compute_mi_cc():
+def test_compute_mi_cc(global_dtype):
```
Done in 354e9c196dd7022c21026afd8d651767b3a8ba7c.

ogrisel commented on sklearn/feature_selection/tests/test_mutual_info.py:
```diff
@@ -43,15 +47,15 @@ def test_compute_mi_cc():
     x, y = Z[:, 0], Z[:, 1]
 
     # Theory and computed values won't be very close, assert that the
     # first figures after decimal point match.
     for n_neighbors in [3, 5, 7]:
         I_computed = _compute_mi(x, y, False, False, n_neighbors)
-        assert_almost_equal(I_computed, I_theory, 1)
+        assert_allclose(I_computed, I_theory, 1)
```
The third argument is now rtol and it should be left to its default value for this test. `atol` should also be left to its 0.0 default value (unless there are zero values in `I_theory`).

```suggestion
        assert_allclose(I_computed, I_theory)
```

jeremiedbb replied:
According to the comment, we should use atol here. rtol=1 means we allow 100% relative error.

ogrisel commented on sklearn/utils/_testing.py:
```diff
@@ -387,6 +395,63 @@ def assert_raise_message(exceptions, message, function, *args, **kwargs):
+    dtypes = [actual.dtype, desired.dtype]
+
+    if rtol is None:
+        rtols = [DTYPE_RELATIVES_TOLERANCES.get(dtype, 1e-9) for dtype in dtypes]
```
Let's keep the numpy default to `1e-7` for any other dtypes to preserve the numpy behavior for those cases.

Actually we can even simplify this to:

```suggestion
        rtols = [1e-4 if dtype == np.float32 or 1e-7 for dtype in dtypes]
```

and then remove the dict-based look-up on `DTYPE_RELATIVES_TOLERANCES`.

We can reintroduce such dict base look-up if we want to generalize this approach to `float16` support but at this point I think this is a [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

ogrisel commented on sklearn/utils/_testing.py:
```diff
@@ -87,6 +87,14 @@
 assert_raises_regexp = assert_raises_regex
 
 
+DTYPE_RELATIVES_TOLERANCES = {
+    np.float32: 1e-4,
+    np.dtype("float32"): 1e-4,
+    np.float64: 1e-7,
+    np.dtype("float64"): 1e-7,
+}
```
I think this can be removed if the suggestion below is considered.

ogrisel commented on sklearn/utils/_testing.py:
```diff
@@ -387,6 +395,63 @@ def assert_raise_message(exceptions, message, function, *args, **kwargs):
+def assert_allclose(
+    actual, desired, rtol=None, atol=0.0, equal_nan=True, err_msg="", verbose=True
+):
+    """
+    Adaptation of numpy.testing.assert_allclose to have tolerances
+    be set based on the input arrays' dtypes.
```
```suggestion
    """dtype-aware variant of numpy.testing.assert_allclose

    This variant introspect the least precise floating point dtype
    in the input argument and automatically sets the relative tolerance
    parameter to 1e-4 float32 and use 1e-7 otherwise (typically float64
    in scikit-learn).

    `atol` is always left to 0.0 by default. It should be adjusted manually
    to an assertion-specific value in case there are null values expected
    in `desired`.

    The aggregate tolerance is ``atol + rtol * abs(desired)``.
```

jeremiedbb commented:
>  Looking at what we have to do to implement our custom assert_allclose and the fact that we won't have a dtype dependent atol, I wonder if it wouldn't be simpler to make the fixture being "global_dtype, rtol, atol" ?

> I do not understand.

nevermind, I thought more about that and I think the custom assert all_close is better.

jeremiedbb commented on doc/computing/parallelism.rst:
```diff
@@ -263,6 +263,14 @@ When this environment variable is set to a non zero value, the tests that need
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When this environment variable is set to '1', the tests using the
+`global_dtype` `pytest.fixture` are also run float32 data.
```
```suggestion
`global_dtype` fixture are also run on float32 data.
```

jeremiedbb commented on sklearn/feature_selection/tests/test_mutual_info.py:
```diff
@@ -43,15 +47,15 @@ def test_compute_mi_cc():
 
     rng = check_random_state(0)
-    Z = rng.multivariate_normal(mean, cov, size=1000)
+    Z = rng.multivariate_normal(mean, cov, size=1000).astype(global_dtype)
```
```suggestion
    Z = rng.multivariate_normal(mean, cov, size=1000).astype(global_dtype, copy=False)
```

ogrisel reviewed (commented):
Could you please add a test in `sklearn.utils.tests.test_testing.py` that checks the behavior of the scikit-learn specific `assert_allclose` variation?

```python
def test_float32_aware_assert_allclose():
    # The relative tolerance for float32 inputs is 1e-4
    assert_allclose(np.array([1. + 2e-5], dtype=np.float32), 1.)
    with pytest.raises(AssertionError):
        assert_allclose(np.array([1. + 2e-4], dtype=np.float32), 1.)

    # The relative tolerance for other inputs is left to 1e-7 as in
    # the original numpy version.
    assert_allclose(np.array([1. + 2e-8], dtype=np.float64), 1.)
    with pytest.raises(AssertionError):
        assert_allclose(np.array([1. + 2e-7], dtype=np.float64), 1.)

    # atol is left to 0.0 by default, even for float32
    with pytest.raises(AssertionError):
        assert_allclose(np.array([1. + 1e-5], dtype=np.float32), 0.)
    assert_allclose(np.array([1. + 1e-5], dtype=np.float32), 0., atol=2e-5)
```

if I am not mistaken (I have not tested).

No need to test the rest of the behavior of `assert_allclose` that is already tested in the numpy test suite.

jeremiedbb commented on sklearn/utils/_testing.py:
```diff
@@ -87,6 +87,14 @@
 
 
+DTYPE_RELATIVES_TOLERANCES = {
+    np.float32: 1e-4,
+    np.dtype("float32"): 1e-4,
```
How does np.dtype("float32") appear ? is it different from np.float32 ?

ogrisel replied:
`np.dtype("float32")` is the attribute of `np.array([0.], dtype=np.float32)`. It is not physically the same as the `np.float32` singleton (`is not`) but it is equal to it (`==`).

jeremiedbb reviewed (commented):
We also need a bit of doc to explain what is the difference between rtol and atol, what are their meaning and when to set atol. Not sure where to put that. Maybe in `contributing.rst` ?

jjerphan committed:
Review comments

Co-authored-by: Jérémie du Boisberranger
<jeremiedbb@users.noreply.github.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jeremiedbb commented on doc/developers/develop.rst:
```diff
@@ -774,3 +774,18 @@ The reason for this setup is reproducibility:
+but you can override via ``rtol``.
+
+When comparing arrays of zero-elements, please do provide a non-zero value for
+the absolute via ``atol``.
```
```suggestion
the absolute tolerance via ``atol``.
```

jeremiedbb commented on sklearn/feature_selection/tests/test_mutual_info.py:
```diff
@@ -47,15 +47,15 @@ def test_compute_mi_cc(global_dtype):
     for n_neighbors in [3, 5, 7]:
         I_computed = _compute_mi(x, y, False, False, n_neighbors)
-        assert_allclose(I_computed, I_theory, 1)
+        assert_allclose(I_computed, I_theory, rtol=1e-1)
```
we should either use atol or rewrite the comment because ``first figures after decimal`` means absolute tolerance

jjerphan replied:
Indeed, I did changes here brainlessly.

jeremiedbb reviewed (approved):
LGTM. Just a last remark from my side.

jjerphan committed:
TST Last changes

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

ogrisel commented on doc/developers/develop.rst:
```diff
@@ -774,3 +774,18 @@ The reason for this setup is reproducibility:
+Numerical assertions in tests
+-----------------------------
+
+When asserting the quasi-equality of arrays, do use
+:func:`sklearn.utils._testing.assert_allclose`.
+
+The relative tolerance is automatically inferred from the provided arrays dtypes,
+but you can override via ``rtol``.
```
```suggestion
When asserting the quasi-equality of arrays of continuous values,
do use :func:`sklearn.utils._testing.assert_allclose`.

The relative tolerance is automatically inferred from the provided arrays
dtypes (for float32 and float64 dtypes in particular) but you can override
via ``rtol``.
```

ogrisel commented on sklearn/utils/tests/test_testing.py:
```diff
@@ -854,3 +855,21 @@ def test_raises():
+    with pytest.raises(AssertionError):
+        assert_allclose(np.array([1.0 + 2e-7], dtype=np.float64), 1.0)
+
+    # atol is left to 0.0 by default, even for float32
+    with pytest.raises(AssertionError):
+        assert_allclose(np.array([1.0 + 1e-5], dtype=np.float32), 0.0)
+    assert_allclose(np.array([1.0 + 1e-5], dtype=np.float32), 1.0, atol=2e-5)
```
Thanks for fixing my suggestion but the following is even closer to what I had in mind (writing almost 0.0 equality checks in tests):

```suggestion
    # atol is left to 0.0 by default, even for float32
    with pytest.raises(AssertionError):
        assert_allclose(np.array([1e-5], dtype=np.float32), 0.0)
    assert_allclose(np.array([1e-5], dtype=np.float32), 0.0, atol=2e-5)
```

ogrisel commented on sklearn/utils/_testing.py:
```diff
@@ -387,6 +387,72 @@ def assert_raise_message(exceptions, message, function, *args, **kwargs):
+        rtols = [1e-4 if dtype == np.float32 else 1e-7 for dtype in dtypes]
+        rtol = max(rtols)
+
+    np_assert_allclose(actual, desired, rtol, atol, equal_nan, err_msg, verbose)
```
Let's call the kwargs, just in case numpy inserts a new kwarg between existing ones in a future version:

```suggestion
    np_assert_allclose(
        actual,
        desired,
        rtol=rtol,
        atol=atol,
        equal_nan=equal_nan,
        err_msg=err_msg,
        verbose=verbose,
    )
```

ogrisel reviewed (approved):
Final round of nitpicks but other than that LGTM!

ogrisel committed:
Apply suggestions from code review

jeremiedbb commented:
Thanks @jjerphan !

jjerphan commented:
I made myself suffer on this one. :upside_down_face:
